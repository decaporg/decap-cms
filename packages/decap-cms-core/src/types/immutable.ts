import { Map as ImmutableMap, List } from 'immutable';

export function isImmutableMap(value: unknown): value is ImmutableMap<string, unknown> {
  return ImmutableMap.isMap(value);
}

export function isImmutableList(value: unknown): value is List<unknown> {
  return List.isList(value);
}

export interface StaticallyTypedRecord<T> {
  get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K];
  set<K extends keyof T, V extends T[K]>(key: K, value: V): StaticallyTypedRecord<T> & T;
  has<K extends keyof T>(key: K): boolean;
  delete<K extends keyof T>(key: K): StaticallyTypedRecord<T>;
  getIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    defaultValue?: V,
  ): T[K1][K2];
  getIn<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    V extends T[K1][K2][K3],
  >(
    keys: [K1, K2, K3],
    defaultValue?: V,
  ): T[K1][K2][K3];
  getIn(keys: string[]): unknown;
  setIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    value: V,
  ): StaticallyTypedRecord<T>;
  setIn(keys: string[], value: unknown): StaticallyTypedRecord<T> & T;
  toJS(): T;
  isEmpty(): boolean;
  some<K extends keyof T>(predicate: (value: T[K], key: K, iter: this) => boolean): boolean;
  mapKeys<K extends keyof T, V>(mapFunc: (key: K, value: StaticallyTypedRecord<T>) => V): V[];
  find<K extends keyof T>(findFunc: (value: T[K]) => boolean): T[K];
  filter<K extends keyof T>(
    predicate: (value: T[K], key: K, iter: this) => boolean,
  ): StaticallyTypedRecord<T>;
  valueSeq<K extends keyof T>(): T[K][] & { toArray: () => T[K][] };
  map<K extends keyof T, V>(
    mapFunc: (value: T[K]) => V,
  ): StaticallyTypedRecord<{ [key: string]: V }>;
  keySeq<K extends keyof T>(): { toArray: () => K[] };
  withMutations(mutator: (mutable: StaticallyTypedRecord<T>) => unknown): StaticallyTypedRecord<T>;
}
