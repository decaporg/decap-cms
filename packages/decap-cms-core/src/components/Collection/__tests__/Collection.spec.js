import React from 'react';
import { render } from '@testing-library/react';
import { fromJS } from 'immutable';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import ConnectedCollection, { Collection } from '../Collection';

jest.mock('../Entries/EntriesCollection', () => 'mock-entries-collection');
jest.mock('../CollectionTop', () => 'mock-collection-top');
jest.mock('../CollectionControls', () => 'mock-collection-controls');
jest.mock('../Sidebar', () => 'mock-sidebar');

const middlewares = [];
const mockStore = configureStore(middlewares);

function renderWithRedux(component, { store } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return render(component, { wrapper: Wrapper });
}

describe('Collection', () => {
  const collection = fromJS({
    name: 'pages',
    sortable_fields: [],
    view_filters: [],
    view_groups: [],
  });
  const props = {
    collections: fromJS([collection]).toOrderedMap(),
    collection,
    collectionName: collection.get('name'),
    t: jest.fn(key => key),
    onSortClick: jest.fn(),
  };

  it('should render with collection without create url', () => {
    const { asFragment } = render(
      <Collection {...props} collection={collection.set('create', false)} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it('should render with collection with create url', () => {
    const { asFragment } = render(
      <Collection {...props} collection={collection.set('create', true)} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with collection with create url and path', () => {
    const { asFragment } = render(
      <Collection {...props} collection={collection.set('create', true)} filterTerm="dir1/dir2" />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render connected component', () => {
    const store = mockStore({
      collections: props.collections,
      entries: fromJS({}),
    });

    const { asFragment } = renderWithRedux(<ConnectedCollection match={{ params: {} }} />, {
      store,
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
