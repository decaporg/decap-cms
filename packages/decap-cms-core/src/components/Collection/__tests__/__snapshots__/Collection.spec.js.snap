// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection should render connected component 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 28px 18px;
}

.emotion-2 {
  padding-left: 280px;
}

<div
    class="emotion-0 emotion-1"
  >
    <mock-sidebar
      collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] }"
      collections="OrderedMap { 0: Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] } }"
      filterterm=""
      searchterm=""
    />
    <main
      class="emotion-2 emotion-3"
    >
      <mock-collection-top
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] }"
        newentryurl=""
      />
      <mock-collection-controls
        filter="Map {}"
        group="Map {}"
        sortablefields=""
        viewfilters=""
        viewgroups=""
      />
      <mock-entries-collection
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] }"
        filterterm=""
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`Collection should render with collection with create url 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 28px 18px;
}

.emotion-2 {
  padding-left: 280px;
}

<div
    class="emotion-0 emotion-1"
  >
    <mock-sidebar
      collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
      collections="OrderedMap { 0: Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] } }"
    />
    <main
      class="emotion-2 emotion-3"
    >
      <mock-collection-top
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
        newentryurl="/collections/pages/new"
      />
      <mock-collection-controls />
      <mock-entries-collection
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`Collection should render with collection with create url and path 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 28px 18px;
}

.emotion-2 {
  padding-left: 280px;
}

<div
    class="emotion-0 emotion-1"
  >
    <mock-sidebar
      collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
      collections="OrderedMap { 0: Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] } }"
      filterterm="dir1/dir2"
    />
    <main
      class="emotion-2 emotion-3"
    >
      <mock-collection-top
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
        newentryurl="/collections/pages/new?path=dir1/dir2"
      />
      <mock-collection-controls />
      <mock-entries-collection
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": true }"
        filterterm="dir1/dir2"
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`Collection should render with collection without create url 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 28px 18px;
}

.emotion-2 {
  padding-left: 280px;
}

<div
    class="emotion-0 emotion-1"
  >
    <mock-sidebar
      collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": false }"
      collections="OrderedMap { 0: Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [] } }"
    />
    <main
      class="emotion-2 emotion-3"
    >
      <mock-collection-top
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": false }"
        newentryurl=""
      />
      <mock-collection-controls />
      <mock-entries-collection
        collection="Map { \\"name\\": \\"pages\\", \\"sortable_fields\\": List [], \\"view_filters\\": List [], \\"view_groups\\": List [], \\"create\\": false }"
      />
    </main>
  </div>
</DocumentFragment>
`;
